File: frontend/lib/domain/entities/chat_message.dart
import 'package:equatable/equatable.dart';

enum MessageRole { user, assistant }

class ChatMessage extends Equatable {
  final String id;
  final MessageRole role;
  final String content;
  final DateTime timestamp;
  final bool isStreaming;

  const ChatMessage({
    required this.id,
    required this.role,
    required this.content,
    required this.timestamp,
    this.isStreaming = false,
  });

  ChatMessage copyWith({
    String? id,
    MessageRole? role,
    String? content,
    DateTime? timestamp,
    bool? isStreaming,
  }) {
    return ChatMessage(
      id: id ?? this.id,
      role: role ?? this.role,
      content: content ?? this.content,
      timestamp: timestamp ?? this.timestamp,
      isStreaming: isStreaming ?? this.isStreaming,
    );
  }

  bool get isUser => role == MessageRole.user;
  bool get isAssistant => role == MessageRole.assistant;

  @override
  List<Object?> get props => [id, role, content, timestamp, isStreaming];
}

File: backend/app/models/chat_message.py
from sqlalchemy import (
    Column,
    BigInteger,
    Enum,
    Text,
    DateTime,
    ForeignKey,
)
from sqlalchemy.dialects.mysql import JSON
from sqlalchemy.sql import func

from app.core.database import Base


class ChatMessage(Base):
    __tablename__ = "chat_message"

    message_id = Column(BigInteger, primary_key=True, autoincrement=True)
    room_id = Column(BigInteger, ForeignKey("chat_room.room_id"), nullable=False)

    role = Column(Enum("user", "assistant"), nullable=False)
    msg_type = Column(Enum("TEXT", "CARD"), nullable=False, server_default="TEXT")

    content = Column(Text, nullable=True)
    payload_json = Column(JSON, nullable=True)

    sent_at = Column(DateTime(timezone=True), server_default=func.now())

    parent_id = Column(BigInteger, nullable=True)
    status = Column(Enum("final", "error"), default="final")
